stdlib/experimental/set_test.go

```diff
@@ -0,0 +1,378 @@
+package experimental_test
+
+import (
+	"sort"
+	"testing"
+
+	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/execute"
+	"github.com/influxdata/flux/execute/executetest"
+	"github.com/influxdata/flux/stdlib/experimental"
+	"github.com/influxdata/flux/values"
+)
+
+// newOrderedObject creates a object with sorted keys order
+func newOrderedObject(vs map[string]values.Value) values.Object {
+	obj := values.NewObjectWithBacking(len(vs))
+	keys := make([]string, 0, len(vs))
+	for k := range vs {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, k := range keys {
+		obj.Set(k, vs[k])
+	}
+	return obj
+}
+
+func TestSet_Process(t *testing.T) {
+	testCases := []struct {
+		name string
+		spec *experimental.SetProcedureSpec
+		data []flux.Table
+		want []*executetest.Table
+	}{
+		{
+			name: "new col",
+			spec: &experimental.SetProcedureSpec{
+				Object: newOrderedObject(map[string]values.Value{
+					"a": values.NewInt(23),
+					"b": values.NewString("hi"),
+				}),
+			},
+			data: []flux.Table{&executetest.Table{
+				ColMeta: []flux.ColMeta{
+					{Label: "_time", Type: flux.TTime},
+					{Label: "_value", Type: flux.TFloat},
+				},
+				Data: [][]interface{}{
+					{execute.Time(1), 2.0},
+					{execute.Time(2), 1.0},
+				},
+			}},
+			want: []*executetest.Table{{
+				ColMeta: []flux.ColMeta{
+					{Label: "_time", Type: flux.TTime},
+					{Label: "_value", Type: flux.TFloat},
+					{Label: "a", Type: flux.TInt},
+					{Label: "b", Type: flux.TString},
+				},
+				Data: [][]interface{}{
+					{execute.Time(1), 2.0, int64(23), "hi"},
+					{execute.Time(2), 1.0, int64(23), "hi"},
+				},
+			}},
+		},
+		{
+			name: "new col with null value",
+			spec: &experimental.SetProcedureSpec{
+				Object: newOrderedObject(map[string]values.Value{
+					"t1": values.NewString("bob"),
+				}),
+			},
+			data: []flux.Table{&executetest.Table{
+				ColMeta: []flux.ColMeta{
+					{Label: "_time", Type: flux.TTime},
+					{Label: "_value", Type: flux.TFloat},
+				},
+				Data: [][]interface{}{
+					{execute.Time(1), nil},
+					{execute.Time(2), 1.0},
+				},
+			}},
+			want: []*executetest.Table{{
+				ColMeta: []flux.ColMeta{
+					{Label: "_time", Type: flux.TTime},
+					{Label: "_value", Type: flux.TFloat},
+					{Label: "t1", Type: flux.TString},
+				},
+				Data: [][]interface{}{
+					{execute.Time(1), nil, "bob"},
+					{execute.Time(2), 1.0, "bob"},
+				},
+			}},
+		},
+		{
+			name: "replace col",
```

Comment:
Perhaps it is a requirement that `set` has both update and extend semantics? But if not, I would like for it to just have extend semantics. That is, I would like for `set` to just extend its input with new columns and not update existing columns.

The reason for this is that we can't infer a principal type for an object when a single operation might update an existing field on an object or extend that object with a new field. These operations have to be distinct if we want to be able to perform inference on them.

I know the current (non-experimental) `set` function does not do this either, but my understanding is that at some point this `set` function will replace our current one, and so I would like the new version to be able to assign the new one a valid type.